terraform/tfschema/customtype.go

```diff
@@ -0,0 +1,230 @@
+limitations under the License.
+*/
+
+package tfschema
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gravitational/protoc-gen-terraform/accessors"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/wrappers"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/trace"
+	schema "github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// SchemaBoolOption represents schema of custom bool value with true default
+func SchemaBoolOption() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeBool,
+		Optional: true,
+		Default:  true,
+	}
+}
+
+// SchemaTraits represents traits schema
+func SchemaTraits() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"key": {
+					Type:     schema.TypeString,
+					Required: true,
+				},
+				"value": {
+					Type:     schema.TypeList,
+					Required: true,
+					Elem: &schema.Schema{
+						Type: schema.TypeString,
+					},
+				},
+			},
+		},
+	}
+}
+
+// SchemaLabels represents traits schema map[string]utils.Strings
+func SchemaLabels() *schema.Schema {
+	return SchemaTraits()
+}
+
+// GetBoolOption unmarshals bool value
+func GetBoolOption(path string, target reflect.Value, meta *accessors.SchemaMeta, sch *schema.Schema, data *schema.ResourceData) error {
+	v := data.Get(path)
+
+	b, ok := v.(bool)
+	if !ok {
+		return trace.Errorf("can not convert %T to bool", v)
+	}
+
+	o := &types.BoolOption{Value: b}
+	target.Set(reflect.ValueOf(o))
+
+	return nil
+}
+
+// getArrayMap gets map of arrays from data to object
+func getArrayMap(
+	obj interface{},
+	sliceT reflect.Type,
+	path string,
+	target reflect.Value,
+	meta *accessors.SchemaMeta,
+	sch *schema.Schema,
+	data *schema.ResourceData,
+) error {
+	l, err := accessors.GetLen(path, data)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if l == 0 {
+		target.Set(reflect.ValueOf(obj))
+		return nil
+	}
+
+	raw := data.Get(path)
+
+	s, ok := raw.(*schema.Set)
+	if !ok {
+		return trace.Errorf("can not convert %T to *schema.Set", raw)
+	}
+
+	for _, i := range s.List() {
+		h := s.F(i)
+
+		k := data.Get(fmt.Sprintf("%v.%v.key", path, h))
+		v := data.Get(fmt.Sprintf("%v.%v.value", path, h))
+
+		vi, ok := v.([]interface{})
+		if !ok {
+			return trace.Errorf("can not convert %T to []interface{}", v)
+		}
+
+		t := reflect.MakeSlice(sliceT, len(vi), len(vi))
+		for i, s := range vi {
+			t.Index(i).Set(reflect.ValueOf(s))
+		}
+
+		reflect.ValueOf(obj).SetMapIndex(reflect.ValueOf(k), t)
+	}
+
+	if s.Len() > 0 {
+		target.Set(reflect.ValueOf(obj))
+	}
+
+	return nil
+}
+
+// GetLabels unmarshals wrapper value map[string]utils.Strings
+func GetLabels(path string, target reflect.Value, meta *accessors.SchemaMeta, sch *schema.Schema, data *schema.ResourceData) error {
+	var s []string = make(utils.Strings, 15)
```

Comment:
Why `15`, here and below? If there's a reason, please move it to a constant with godoc explaining it.

Reply:
No, sorry, replaced it with proper reflect statement, removed make.